File: backend/app/schemas/end_card.py
"""Pydantic schemas for the mandatory end card."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from app.models.end_card import (
    MAX_END_CARD_SECONDS,
    MIN_END_CARD_SECONDS,
    EndCardAudioMode,
    EndCardLayout,
)


class EndCardSettingsUpdate(BaseModel):
    """Partial update for the global or per-project end card configuration."""

    layout: EndCardLayout | None = None
    duration_seconds: float | None = Field(
        default=None, ge=MIN_END_CARD_SECONDS, le=MAX_END_CARD_SECONDS
    )
    fade_in_ms: int | None = Field(default=None, ge=0, le=3000)
    audio_fade_out_ms: int | None = Field(default=None, ge=0, le=5000)
    audio_mode: EndCardAudioMode | None = None
    music_volume: float | None = Field(default=None, ge=0.0, le=1.0)
    url_text: str | None = Field(default=None, max_length=500)
    show_qr: bool | None = None
    qr_url: str | None = Field(default=None, max_length=500)
    channel_handle: str | None = Field(default=None, max_length=200)
    custom_message: str | None = Field(default=None, max_length=500)


class EndCardSettingsResponse(BaseModel):
    """Effective end card configuration, with its provenance."""

    model_config = ConfigDict(from_attributes=True)

    layout: EndCardLayout
    duration_seconds: float
    fade_in_ms: int
    audio_fade_out_ms: int
    audio_mode: EndCardAudioMode
    music_filename: str | None
    music_volume: float
    logo_filename: str | None
    url_text: str | None
    show_qr: bool
    qr_url: str | None
    channel_handle: str | None
    custom_message: str | None
    # False when the project inherits the global configuration.
    is_project_override: bool
    # The end card cannot be disabled; surfaced so the UI can say so.
    is_mandatory: bool = True
    min_duration_seconds: float = MIN_END_CARD_SECONDS
    max_duration_seconds: float = MAX_END_CARD_SECONDS


class EndCardLayoutInfo(BaseModel):
    """A selectable design."""

    id: EndCardLayout
    label: str
    description: str
    needs_cover: bool


class EndCardLayoutListResponse(BaseModel):
    items: list[EndCardLayoutInfo]

File: frontend/src/types/endCard.ts
export type EndCardLayout = 'cover_full' | 'cover_card' | 'minimal';
export type EndCardAudioMode = 'silence' | 'continue_with_fade' | 'local_music';
export type EndCardMessagePosition = 'top' | 'center' | 'bottom';

export interface EndCardLayoutInfo {
  id: EndCardLayout;
  label: string;
  description: string;
  needs_cover: boolean;
}

export interface EndCardSettings {
  layout: EndCardLayout;
  duration_seconds: number;
  fade_in_ms: number;
  audio_fade_out_ms: number;
  audio_mode: EndCardAudioMode;
  music_filename: string | null;
  music_volume: number;
  logo_filename: string | null;
  url_text: string | null;
  show_qr: boolean;
  qr_url: string | null;
  channel_handle: string | null;
  custom_message: string | null;
  message_position: EndCardMessagePosition;
  /** False when the project inherits the global configuration. */
  is_project_override: boolean;
  /** The end card cannot be turned off. */
  is_mandatory: boolean;
  min_duration_seconds: number;
  max_duration_seconds: number;
}

export type EndCardSettingsPayload = Partial<
  Pick<
    EndCardSettings,
    | 'layout'
    | 'duration_seconds'
    | 'fade_in_ms'
    | 'audio_fade_out_ms'
    | 'audio_mode'
    | 'music_volume'
    | 'url_text'
    | 'show_qr'
    | 'qr_url'
    | 'channel_handle'
    | 'custom_message'
    | 'message_position'
  >
>;
